src/Polysemy/RandomFu.hs

```diff
@@ -0,0 +1,110 @@
+
+
+Also, its presence in the Polysemy Union @r@,
+gives @Sem r@ a "MonadRandom" instance (from "random-fu"). This is an Orphan instance.
+-}
+
+module Polysemy.RandomFu
+  (
+    -- * Effect
+    Random (..)
+
+    -- * Actions
+  , sampleRVar
+  , getRandomPrim
+  , sampleDist
+
+  -- * Interpretations
+  , runRandomSource
+  , runRandomIO
+  , runRandomIOPureMT
+  )
+where
+
+import           Polysemy                      
+
+import           Data.IORef                     ( newIORef )
+import qualified Data.Random                   as R
+import qualified Data.Random.Source            as R
+import           Data.Random.Source.Std         (StdRandom(..))
+import qualified Data.Random.Internal.Source   as R
+import qualified Data.Random.Source.PureMT     as R
+import qualified System.Random                 as SR
+import           Control.Monad.IO.Class         ( MonadIO(..) )
+
+------------------------------------------------------------------------------
+{- | An effect capable of sampling from a "random-fu" RVar or generating a
+single random-variate of any type, @t@ with a
+@Data.Random.Prim t@ constructor, currently one of @Word8@, @Word16@,
+@Word32@, @Word64@, @Double@ or N-byte integer.
+-}
+data Random m r where
+  SampleRVar ::  R.RVar t -> Random m t
+  GetRandomPrim :: R.Prim t -> Random m t
+
+makeSem ''Random
+
+------------------------------------------------------------------------------
+-- | use the 'Random` effect to sample from a "random-fu" @Distribution@.
+sampleDist
+  :: (Member Random r, R.Distribution d t) => d t -> Sem r t
+sampleDist = sampleRVar . R.rvar
+
+------------------------------------------------------------------------------
+-- | Run a 'Random' effect using a given 'R.RandomSource'
+runRandomSource
+  :: forall s r a
+   . R.RandomSource (Sem r) s
+  => s
+  -> Sem (Random ': r) a
+  -> Sem r a
+runRandomSource source = interpret f
+ where
+  f :: forall m x . (Random m x -> Sem r x)
+  f r = case r of
+    SampleRVar    rv -> R.runRVar (R.sample rv) source
+    GetRandomPrim pt -> R.runRVar (R.getRandomPrim pt) source
+{-# INLINE runRandomSource #-}
+
+------------------------------------------------------------------------------
+-- | Run a 'Random` effect by using the default "random-fu" 'IO' source
+runRandomIO
+  :: forall r a
+   . MonadIO (Sem r)
+  => Sem (Random ': r) a
+  -> Sem r a
+runRandomIO = interpret f
+ where
+  f :: forall m x . (Random m x -> Sem r x)
+  f r = case r of
+    SampleRVar    rv -> liftIO $ R.sample rv
+    GetRandomPrim pt -> liftIO $ R.getRandomPrim pt
+{-# INLINE runRandomIO #-}
+
+------------------------------------------------------------------------------
+-- | Run in 'IO', using the given 'R.PureMT' source, stored in an 'IORef'
+runRandomIOPureMT
+  :: MonadIO (Sem r)
+  => R.PureMT
+  -> Sem (Random ': r) a
+  -> Sem r a
+runRandomIOPureMT source re =
+  liftIO (newIORef source) >>= flip runRandomSource re
```

Comment:
To confirm my understanding, an `IORef PureMT` is a `RandomSource`?

Reply:
Yes.  Random-fu is a little confusing to me.  There are sources ```s```, where ```RandomSource m s``` means that ```s``` can be "sampled" in ```m```.  There is also ```MonadRandom m``` which sort of implies that there exists a source such that it can be sampled in ```m```.  Users, in the sense of functions that would like to be able to sample random-variables, don't need to know about the source.  But when you run the thing, you do.  I think all of this is just to generalize over ```StdGen``` and ```PureMT``` and possible future generators.  But I'm not sure.

Reply:
Random fu is confusing to everyone :)

Is the `MonadRandom` bit necessary?